WebXRManager: Added depth sensing support

cabanier opened the pull request:
First pass at adding support for WebXR GPU based depth sensing.
Also updated the XR samples to request depth.

We can tweak shaders and parameters later.

*This contribution is funded by [Meta](https://meta.com)*

cabanier committed:
add support for depth sensing

CodyJasonBennett commented on src/renderers/webgl/WebGLProgram.js:
```diff
@@ -510,6 +510,8 @@ function WebGLProgram( renderer, cacheKey, parameters, bindingStates ) {
 			parameters.useFog && parameters.fog ? '#define USE_FOG' : '',
 			parameters.useFog && parameters.fogExp2 ? '#define FOG_EXP2' : '',
 
+			parameters.occlusion ? '#define USE_OCCLUSION' : '#define USE_OCCLUSION',
+
```
Both sides are equal; the right-hand side should be `''`.

cabanier replied:
I know :-)
That is why it's WIP and still a draft PR

cabanier replied:
@CodyJasonBennett fixed!

hybridherbst commented:
Super cool!

Out of curiosity, have you done a perf/quality comparison between:
- passing depth into each material as in this PR
- drawing a fullscreen pass that renders depth only and sets z buffer values so others test against  that
- doing that z writing from e.g. the background material (so it stays one pass)

I’d think writing the z buffer directly loses the fuzziness but should be faster (can use early z discard, no extra texture fetch per fragment, etc)

cabanier commented:
> Super cool!
> 
> Out of curiosity, have you done a perf/quality comparison between:
> 
> * passing depth into each material as in this PR
> * drawing a fullscreen pass that renders depth only and sets z buffer values so others test against  that, for example right from the background material
> 
> I’d think writing the z buffer directly loses the fuzziness but should be faster (can use early z discard, no extra texture fetch per fragment, etc)

I have not done a perf comparison. This way seems to integrate nicely with three, although it's a bit special since it's a renderer feature.
We can revisit more optimal approaches later. I can also investigate how this is done in unity and unreal.

mrdoob commented:
@hybridherbst

> drawing a fullscreen pass that renders depth only and sets z buffer values so others test against that, for example right from the background material

I feel like this approach is much more straight forward and easier to maintain, regardless of the performance implications.

I think I would try implementing this approach first before modifying all the shaders for all materials.

cabanier commented on src/renderers/webxr/WebXRManager.js:
```diff
@@ -730,6 +739,32 @@ class WebXRManager extends EventDispatcher {
+
+						if (( depthData.depthNear != session.renderState.depthNear ) || ( depthData.depthFar != session.renderState.depthFar )) {
+
+							session.updateRenderState( {depthNear: depthData.depthNear, depthFar: ( depthData.depthFar > 10000 ? 10000 : depthData.depthFar ) } );
```
this seems like a WebXR bug

toji replied:
Spec looks like it allows infinity:

> If activeState’s [depthFar](https://immersive-web.github.io/webxr/#dom-xrrenderstate-depthfar) is greater than session’s [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane) set activeState’s [depthFar](https://immersive-web.github.io/webxr/#dom-xrrenderstate-depthfar) to session’s [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane).

And

> The [maximum far clip plane](https://immersive-web.github.io/webxr/#maximum-far-clip-plane) SHOULD be greater than 1000.0 (and MAY be infinite).

So if it's not working then it's a Chromium implementation issue. Idea is to allow `depthFar: Infinity`, right?

cabanier replied:
> So if it's not working then it's a Chromium implementation issue. Idea is to allow `depthFar: Infinity`, right?

Yes. If I pass infinity, I get an error saying the value is non-finite

cabanier commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1600,7 +1615,7 @@ class WebGLRenderer {
 			const lightsStateVersion = lights.state.version;
 
-			const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, object );
+			const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, object, _occlusion );
```
@mrdoob is there a better way than passing in a renderer variable?

cabanier commented:
@mrdoob @Mugen87 would you mind taking a look?

snowzurfer commented on src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:
```diff
@@ -0,0 +1,26 @@
+    if (gl_FragColor.a > 0.0) {
+        int arrayIndex = 0;
+        vec2 depthUv;
+        if (gl_FragCoord.x>=depthWidth) {
```
is this checking which eye to read the texture for?

cabanier replied:
Yes. I considered passing it in as a value but that required deeper changes in the renderer.

snowzurfer commented:
fwiw it looks like on Unity, at least for custom shaders, [they suggest modifying the shaders and  calculating the occlusion in the fragment shader](https://github.com/oculus-samples/Unity-DepthAPI/tree/main#step-4-calculate-occlusions-in-fragment-shader):
```
half4 frag(v2f i) {
   UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(i);

   // this is something your shader will return without occlusions
   half4 fragmentShaderResult = someColor;

   // calculate UV for the depth texture lookup for occlusions
   float2 uv = i.positionNDC.xy / i.positionNDC.w;

   // pass UV and the current depth of the texel
   float occlusionValue = CalculateEnvironmentDepthOcclusion(uv, i.vertex.z);

   // consider early rejection to not write to depth if it's an opaque shader
   if (occlusionValue < 0.01) {
       discard;
   }

   // premultiply color and alpha by occlusion value
   // when it's 1 - color is not affected - virtual covers real
   // when it's 0 - texel is invisible - virtual is under real
   // when it's in between - texel is semi transparent
   fragmentShaderResult *= occlusionValue;

   return result;
}
```

cabanier commented:
> fwiw it looks like on Unity, at least for custom shaders, [they suggest modifying the shaders and calculating the occlusion in the fragment shader](https://github.com/oculus-samples/Unity-DepthAPI/tree/main#step-4-calculate-occlusions-in-fragment-shader)

We can play with the shaders in follow up patches. It might even be interesting to vary which depth values are honored, or provide a color to overlay real world objects

gkjohnson committed:
BatchedMesh: cleanup, add maxGeometryCount member (#27231)

* Linting, add maxGeometryCount to BatchedMesh, remove undocumented functions

* BatchedMesh: Update documentation

Mugen87 committed:
AudioLoader: Fix uncaught DOMException in promise. (#27237)

Mugen87 committed:
Updated builds.

sunag committed:
Rename .colorSpace -> .colorSpaced (#27238)

sunag committed:
TSL: Introduction to fragmentNode (#27239)

* Introduction to fragmentNode

* revision

sunag committed:
WebGPURenderer: Fix .clear() using RenderTarget and revisions (#27233)

* Fix .clear() using RenderTarget

* webgpu_rtt: improve a little

* cleanup

* cleanup

* revisions

* revision

* wip flipY

* update example

* cleanup

sunag committed:
Cleanup from https://github.com/mrdoob/three.js/pull/27233

sunag committed:
WebGPURenderer: Depth Pixel & Logarithmic Depth Buffer (#27243)

* WebGPURenderer: Depth Pixel & Logarithmic Depth Buffer

* Update puppeteer.js

gkjohnson committed:
BatchedMesh: Add pure annotations, fix camera position for sorting (#27236)

* Add pure annotations, fix camera position for sorting

* Update annotations

sunag committed:
NodeMaterial: Prioritize .depthNode (#27244)

sunag committed:
Cleanup from https://github.com/mrdoob/three.js/pull/27243

renovate[bot] committed:
Update dependency rollup-plugin-visualizer to v5.9.3 (#27248)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

RenaudRohlinger committed:
WebGLArrayRenderTarget: Missing options in constructor (#27249)

* missing options constructor

* cleanup format

Mugen87 committed:
Docs: Improve render target pages. (#27253)

* Docs: Improve render target pages.

* Docs: Clean up.

Mugen87 committed:
Docs: Improve `WebGLCubeRenderTarget` page. (#27254)

RenaudRohlinger committed:
Object3D: Add onBeforeShadow and onAfterShadow callbacks (#27250)

* added onbeforeShadow and onAfterShadow

* add missing methods

Mugen87 committed:
Docs: Improve `Object3D` page. (#27257)

gkjohnson committed:
BatchedMesh: Add sorting and frustum culling for shadows (#27258)

Mugen87 committed:
Updated builds.

Mugen87 committed:
Examples: Clean up. (#27262)

renovate[bot] committed:
Update github/codeql-action digest to 407ffaf (#27247)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

aardgoose committed:
read tex async (#27006)

Co-authored-by: aardgoose <angus.sawyer@email.com>

Mugen87 committed:
USDZExporter: Remove obsolete code in `buildVector2Array()`. (#27264)

mrdoob committed:
Object3D: Clean up.

mrdoob commented:
@cabanier Is this the latest spec: https://www.w3.org/TR/webxr-depth-sensing-1/

Also, did you see [this feedback](https://github.com/mrdoob/three.js/pull/27154#issuecomment-1803354732
)?

Mugen87 committed:
OrbitControls: Remove useless assignment. (#27266)

mrdoob committed:
Editor: Unified Viewport.Camera and Viewport.Shading,

linbingquan committed:
Examples: Clean up. (#27267)

mrdoob committed:
Editor: Moved grid and helpers settings to Viewport.Controls.

mrdoob committed:
Editor: Fixed helpers checkbox also hiding the grid.

mrdoob committed:
Editor: Clean up.

mrdoob committed:
Editor: Removed IndexedDB prefixes.

mrdoob commented:
> Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?

That's right. We render a quad with the depth data first and then OpenGL does the rest.
No need to modify any built-in shader.

> If so, that would destroy performance on headsets as it would introduce a depth resolve.

Do you mind elaborating this point?

cabanier commented:
> > Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?
> 
> That's right. We render a quad with the depth data first and then OpenGL does the rest. No need to modify any built-in shader.

How would we use the quad with the depth data? Is there a way to do a depth compare between 2 textures? Note that the resolutions of the depth buffer and real world depth are different.

> > If so, that would destroy performance on headsets as it would introduce a depth resolve.
> 
> Do you mind elaborating this point?

Because we're rendering to a multisampled framebuffer, the depth buffer is very big and only needed while we're resolving the pixels. The resolved depth buffer is never used by the system so we make sure that it's invalidated at the very end of the render pass. That way it is not flushed to main memory which saves a lot of GPU time.

cabanier commented:
> > Yes. Maybe I misunderstood it. Do you want the approach where we render to depth first?
> 
> That's right. We render a quad with the depth data first and then OpenGL does the rest. No need to modify any built-in shader.

@mrdoob if you can point me to how I could implement this, I will give it a shot.

linbingquan committed:
src: Clean up. (#27272)

mrdoob commented:
On my Quest 3 with Chrome 118 this sample doesn't work (clicking in START AR does nothing):
https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html

Is there a flag I need to enable?

cabanier commented:
> On my Quest 3 with Chrome 118 this sample doesn't work (clicking on START AR does nothing): https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html
> 
> Is there a flag I need to enable?

Are you on an [early access OS release](https://www.meta.com/en-gb/help/quest/articles/headsets-and-accessories/using-your-headset/quest-ptc/)? You need v60. 
If so, go to `chrome://flags`, enable `WebXR experiments` and restart the browser.

cabanier commented:
> > On my Quest 3 with Chrome 118 this sample doesn't work (clicking on START AR does nothing): https://immersive-web.github.io/webxr-samples/proposals/phone-ar-depth-gpu.html
> > Is there a flag I need to enable?
> 

FYI even with the flag, that sample won't work because it uses the CPU version of depth sensing

Mugen87 committed:
Docs: Improve `LineDashedMaterial` page. (#27276)

daniele-pelagatti committed:
bump @monogrid/gainmap-js to 3.0.0 that fixes problems with cubemap generation using an equirectangular rendertarget with mipmaps (#27277)

Mugen87 committed:
Docs: Improve `LineDashedMaterial` page. (#27278)

sunag committed:
WebGPURenderer: Fix strip index format (#27279)

mrdoob committed:
Editor: Fixed UINumber mobile UX.

mrdoob commented:
Ah ah I see...

Unfortunately I do not have the bandwidth to figure out how the early access stuff works.

Lets hold this until v60 is out 👍

gkjohnson committed:
Adjust the docs on opacity for transmission (#27281)

Mugen87 committed:
Docs: Clean up. (#27282)

RenaudRohlinger committed:
WebGLUniforms: `sampler2DShadow` support (#27252)

* add support for sampler2DShadow in uniforms

* Update WebGLUniforms.js

* use gl.SAMPLER_2D_SHADOW instead

* Update WebGLUniforms.js

---------

Co-authored-by: mrdoob <info@mrdoob.com>

Mugen87 committed:
GLTFExporter: Restore multi material support for non-indexed geometries. (#27268)

RenaudRohlinger committed:
fix safari offscreen canvas without webgl context support (#27283)

mrdoob committed:
Updated builds.

mrdoob committed:
Examples: Improved webgl_loader_texture_hdrjpg.

mrdoob committed:
Examples: Removed minimap from webgpu_tsl_transpiler.

mrdoob committed:
Examples: Clean up.

mrdoob committed:
WebGPURenderer: Added webgpu string in data-engine.

mrdoob committed:
r159

RenaudRohlinger committed:
WebGLUniformsGroups: Correct 3x3 to 3x4 Matrix Conversion in Uniform Buffer Object (#27284)

* fix matrix 3x4 conversion

* fix matrix 3x4 conversion

* assign 0 to 7 and 11

* Update WebGLUniformsGroups.js

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

sunag committed:
WebGPU Compute Rain Example: Fix RTT coord Y (#27286)

Mugen87 committed:
Editor: Fix multi-material support. (#27265)

ianpetrarca committed:
Update 3DMLoader Docs and Example CDN Script (#27289)

* Update 3DMLoader.html

Update the CDN script used in the 3DMLoader in this example.

* Update 3DMLoader.html

Update top HTML content with new CDN URL.

* Update 3DMLoader.html

Update cdn scripts

* Update webgl_loader_3dm.html

Update CDN URL

* Update 3DMLoader.html

Update NPM

WestLangley committed:
Clarify warning messages (#27288)

paulmasson committed:
Update REVISION (#27287)

linbingquan committed:
upgrade rollup to `^4.6.0` (#27275)

RenaudRohlinger committed:
WebGLUniformsGroups: UBO Boolean support and fix cache on number (#27285)

* add boolean support and fix number cache

* fix const assignment

* good fix assignment to local var

* Update WebGLUniformsGroups.js

Improve comment.

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Update package-lock.json

Mugen87 committed:
Updated builds.

renovate[bot] committed:
Update devDependencies (non-major) (#27297)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

puxiao committed:
[Doc/zh] Update Quaternion.html (#27298)

RenaudRohlinger committed:
WebGLUniformsGroups: Support for array of Uniform Buffer Object (#27293)

* ubo array support

* dpr to 1

* update ubo

* Update webgl2_ubo_arrays.html

Make example code more consistent to other demos.

* Update WebGLUniformsGroups.js

Clean up.

* feedbacks

* checkout build folder from dev

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Docs: Improve `SkinnedMesh` page. (#27303)

Benjythebee committed:
DracoLoader: Fix not calling `onError()`. (#27306)

* fix onError load

* onError

* LinearSRGBColorSpace instead of undefined

* default onError = () => {}; call onError to cleanup code;

Benjythebee committed:
DracoLoader: remove extra redundant code (#27308)

* fix onError load

* onError

* LinearSRGBColorSpace instead of undefined

* default onError = () => {}; call onError to cleanup code;

* dracoloader: remove redundant catch

RenaudRohlinger committed:
fix level mipmap and support compressed tex (#27313)

gkjohnson committed:
Only use float textures if linear interpolation is supported (#27315)

Mugen87 committed:
Updated builds.

Rabbid76 committed:
Examples: GTAOPass - Ground Truth Ambient Occlusion (#27317)

Mugen87 committed:
Examples: Clean up (#27318)

* GTAOShader: Clean up.

* Example: Clean up.

Mugen87 committed:
Examples: Simplify composer setup in GTAO demo. (#27319)

* Examples: Simplify composer setup in GTAO demo.

* Examples: Update screenshots.

* Examples: More clean up.

Mugen87 committed:
Examples: Improve AO shaders. (#27320)

* GTAOPass: Use higher default depth precision.

* GTAOShader: Try highp sampler2D.

* Examples: Define high sampler precision for AO shaders.

* Examples: Restrict highp samplers to depth only.

Mugen87 committed:
Examples: Clean up. (#27321)

sunag committed:
TSL: Fix texture_depth_2d in wgslFn (#27323)

Methuselah96 committed:
Fix typo in webgl2_ubo_arrays example (#27322)

sheepdragon committed:
Fixed: Flow reports an error in multiple materials (#27302) (#27324)

Co-authored-by: shiyanglong <shiyanglong@hixianchang.com>

gkjohnson committed:
Triangle: Return `null` if the triangle is degenerated.  (#27311)

* BatchedMesh: Add sorting and frustum culling for shadows

* Triangle.getBarycoord: throw error if a degenerate triangle is used

* Return null, instead

* Fix test

* Fix containsPoint

Mugen87 committed:
WebGLShadowMap: Dispose of unique depth/distance materials. (#27326)

* WebGLShadowMap: Dispose of unique depth/distance materials.

* Update WebGLShadowMap.js

Fix comment.

* Update WebGLShadowMap.js

Mugen87 committed:
Editor: Add missing `manager` to `GLTFLoader`. (#27336)

Mugen87 committed:
Editor: Fix loading zip files with `.glTF`. (#27337)

sunag committed:
StorageTexture: Mipmaps (#27332)

* StorageTexture: Mipmaps

* revision

* use texture.generateMipmaps

Archimagus committed:
ImageBitmapLoader: Cache promises to deduplicate requests. (#27270)

* Fix for dispose error
sampleRenderTarget is being deleted and set to null by super.dispose();
the check for undefined was incorrect causing an attempt to call dispose on a null object.

* Removed disposal of sampleRenderTarget entierly

* Fix for setting path not affecting GLTF Sub Assets correctly.

* Fixed Comment

* Update GLTFLoader.js

Fix code style.

* Added additional caching to ImageBitmapLoader
So that if multiple calls to load the same image come in,
before the first one is loaded it will still only load it once.

* Lint

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

josephrocca committed:
Add Jolt to physics list (#27335)

* Add Jolt to physics list

* Add Jolt to physics (ar)

* Add Jolt to Physics (fr)

* Add jolt to physics (it)

* Add jolt to physics (ja)

* Update Libraries-and-Plugins.html

* Update Libraries-and-Plugins.html

* Add jolt to physics (pt-br)

* Add jolt to physics (ru)

* Add jolt to physics (zh)

sunag committed:
WebGPURenderer: MaterialX (#27294)

* MathNode: Add `bitcast()`

* WGSLNodeBuilder: Add `greaterThan` for vec3

* WGSLNodeBuilder: XOR

* NodeBuilder: .buildFunctionNode() and getFunctionOperator()

* TSL: Function Overloading

* SplitNode: Fix expected output if used uvec or ivec

* TSL Transpiler: Improvements and fixes

* TSL: Fix expected output if used a.assign( b ), returns a

* FIX & TODO: Check reasons

* Fix clearcoat op sequence

* WebGPURenderer: MaterialX Support

* add webgpu materialx examples

* fix backend name

* unknown diff screenshot problem

* testing e2e response

Mugen87 committed:
Core: Remove deprecated code. (#27341)

* Core: Remove deprecated code.

* Tests: Clean up.

Mugen87 committed:
WebGLTextures: Directly evaluate `capabilities`. (#27340)

gkjohnson committed:
Triangle.getInterpolation: Return null if triangle is degenerate (#27331)

* Return null if triangle is degenerate

* Update docs

* Ensure we support for vector4 and vector2

* Update Triangle.html

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Rabbid76 committed:
Examples: GTAOPass - remove unused uniform variable and parameter (#27348)

sunag committed:
Core: Remove some defines unused (#27346)

renovate[bot] committed:
Update devDependencies (non-major) (#27351)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action digest to c0d1daa (#27350)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

gkjohnson committed:
LineMaterial: Fix clipping at edge artifacts (#27349)

* Fix containing cube

* Sceenshot update

OndrejSpanel committed:
Fix: CSM lights when shadow maps are disabled (#27358)

* Fix indentation

* Allow shadows to be disabled from GUI

* Fix: when shadows are disabled, light contribution still needs to be considered

cabanier committed:
Add support for offerSession (#27359)

hybridherbst committed:
feat: add normalNode and emissiveNode support for MaterialXLoader (#27356)

Mugen87 committed:
Quaternion: Add missing call of `_onChangeCallback()` to `fromBufferAttribute()`. (#27363)

Mugen87 committed:
Quaternion: Add default parameter to `setFromEuler()`. (#27364)

Mugen87 committed:
Quaternion: Remove redundant `_onChangeCallback` in `slerp()`. (#27365)

hybridherbst committed:
Nodes: Log warnings instead of throwing on redefinitions of Node types (#27357)

This is the same behaviour as three.js itself has when it's imported multiple times.

cabanier committed:
Call offerSession at session end time (#27368)

Rabbid76 committed:
GTAOPass: distance dependent horizon search (#27370)

linbingquan committed:
Docs: Fixed import error (#27372)

aardgoose committed:
WebGPURenderer: support getArrayBufferAsync() for WebGL backend (#27355)

* read buffer async

* return ArrayBuffer

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

RenaudRohlinger committed:
WebGLUniformsGroups: Fix and improve UBOs array caching system (#27374)

* fix and improve UBOs array caching system

* simplify cache check

CodyJasonBennett committed:
Nodes: Mark directory as effectful for Webpack/Vite. (#27376)

cabanier commented:
> Lets hold this until v60 is out 👍

v60 is now out. Browser 30.3 is also out with experimental support. Go to `chrome://flags`, enable `WebXR experiments` and restart the browser.

I'm still unconvinced that a depth prepass is a practical solution.  See  https://keaukraine.medium.com/efficient-rendering-of-soft-particles-on-mobile-gpus-9beb856fcfbc or https://developer.arm.com/documentation/101897/0301/Optimizing-application-logic/Avoid-using-depth-prepasses on why that is

sunag committed:
WebGPURenderer: PostProcessing + GaussianBlurNode + QuadMesh (#27369)

* WebGPURenderer: PostProcessing + GaussianBlurNode

* update example

* revision

* revision

* Fix RTT & Framebuffer flipY

* Fix multi-scene backgroundNode

* fixes

* new webgpu_portal example

* fix title

* cleanup

* cleanup

* adjustments

* Added QuadMesh

* PostProcessing just for WebGPUBackend for now

* portal update

* error message

* cleanup

* using quad texture

* Fix flip RTT & DepthNode after QuadMesh

* update to QuadMesh

* Update webgpu_depth_texture.jpg

* update `webgpu_instance_uniform` example

jeffbeene committed:
USDZExporter: export from compressed texture data #23321 (#27382)

* USDZExporter: export from compressed texture data #23321

* add screenshot, update files.json

* Update USDZExporter.js

Clean up.

* remove usdz compressed example

* Update USDZExporter.js

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

RenaudRohlinger committed:
WebGLRenderer: Add `WEBGL_clip_cull_distance` support. (#27371)

* init WEBGL_clip_cull_distance support

* remove unecessary conditions webgl2

* try fix directx

* feedbacks

* remove unecessary code

* Update webgl_clipculldistance.html

Style clean up.

* Update WebGLExtensions.js

* Update webgl_clipculldistance.html

* replace material with shader

* remove unused time const

* Update webgl_clipculldistance.html

Simplify example.

* Update clipping example.

* Revert builds.

* ShaderMaterial: Update `extensions` and docs.

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Updated builds.

jeffbeene committed:
USDZExporter: only include required modules #27382 (#27383)

sciecode committed:
OrbitControls: introduce zoom-to-cursor on touch events (#27384)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

Mugen87 committed:
Added THREE.Timer. (#17912)

* Added THREE.Timer.

* Timer: Make now() private.

* Timer: Fix typo.

* Timer: Shorten method and variable names.

* Timer: Remove js and add d.ts file.

* Timer: Transform to class.

* Update webgl_morphtargets_sphere.html

* Delete Timer.d.ts

* Update webgl_morphtargets_sphere.html

* Timer: Add optional timestamp parameter to update().

* Timer: Fix multiple instance usage.

* Timer: Clean up.

Mugen87 committed:
Update Timer.js

Add description.

KTRosenberg commented:
Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.

renovate[bot] committed:
Update github/codeql-action digest to 03e7845 (#27389)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v3 (#27392)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update actions/upload-artifact action to v4 (#27391)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

cabanier commented:
> Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.

No, there are no `adb` flags; just the `WebXR experiments` in `chrome://flags`

KTRosenberg commented:
> > Confirming, this sample does NOT work as expected on Quest 3 still? I’m looking forward to trying the equivalent of Oculus’s Unity sample if at all possible, but I’m not sure whether the functionality is exposed. You need to enable a flag in adb, possibly every so often since I noticed it gets disabled again after a while.
> 
> No, there are no `adb` flags; just the `WebXR experiments` in `chrome://flags`

I might have misunderstood your comment above here: https://github.com/mrdoob/three.js/pull/27154#issuecomment-1831447301

This was just referring to the old example for some Android phones.

So the new example here should work on Quest 3?

cabanier commented:
> This was just referring to the old example for some Android phones.
> So the new example here should work on Quest 3?

Yes, it's only for Quest 3. I *think* it should be easy to make it work for android but I don't have a device to test it on.

renovate[bot] committed:
Update devDependencies (non-major) (#27390)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Mugen87 committed:
Docs: Add page for `Timer`. (#27394)

Mugen87 committed:
Sphere: Add `isSphere` flag. (#27395)

sunag committed:
WebGPURenderer: Fix the use of same Material with different Skeleton (#27398)

sunag committed:
WebGPURenderer: Backdrop Water Example (#27397)

* fix triplanar with dynamic mesh

* add example: `webgpu_backdrop_water`

* fix names

* cleanup

* cleanup

* revision

* Update puppeteer.js

* revision

* revision

* caustic

* revision

* revision

* Update webgpu_backdrop_water.html

* update shadow

sunag committed:
WebGPURenderer: Compute Snow Example (#27400)

* NodeMaterial: Check toneMapped=true

* Renderer: overrideMaterial compatible with positionNode

* GaussianBlurNode: Added resolution

* add `webgpu_compute_particles_snow` example

* update

* fix scale offset collision

* update initial pos y

gkjohnson committed:
WebGLRenderer: Add support for AgX Tone Mapping (#27366)

* Add support for AGX tone mapping

* Update tonemapping example

* Update test

* AGX -> AgX

* glsl lint, suffixes fix

* Adjust tone mapping constant values

* Update src/renderers/shaders/ShaderChunk/tonemapping_pars_fragment.glsl.js

Co-authored-by: Don McCurdy <dm@donmccurdy.com>

* Update src/renderers/shaders/ShaderChunk/tonemapping_pars_fragment.glsl.js

Co-authored-by: Don McCurdy <dm@donmccurdy.com>

* Update OutputPass.js

Add `AgX` support.

* Update OutputShader.js

Add `AgX` support.

* Clamp the values to be > 0.0

* Add rec2020 matrices

* Fix rec2020 conversion matrices

* Update implementation to be based on filament using rec 2020 color space

* Spaces -> tabs

* linting

* Code cleanup

* Support WebGL1, rearrange

* Comments update

* Remove redundant clamp

---------

Co-authored-by: Don McCurdy <dm@donmccurdy.com>
Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Updated builds.

sunag committed:
WebGPURenderer: Screenshots (#27402)

hybridherbst committed:
Nodes: use relative references inside examples/jsm (#27381)

Fixes https://github.com/mrdoob/three.js/issues/27354

gkjohnson committed:
Add AgX tone mapping constant (#27403)

sunag committed:
Examples: Improve performance - Compute Snow and Backdrop Water (#27404)

* less particles for better performance

* cleanup

* 60fps on M1

SADYX committed:
Docs: Improve Chinese translation. (#27408)

Mugen87 committed:
Docs: Clean up. (#27409)

aardgoose committed:
WebGPURenderer: bug fixes for WebGLBackend (#27410)

Co-authored-by: aardgoose <angus.sawyer@email.com>

cabanier commented:
@mrdoob any thoughts? Did you try this out?

hybridherbst commented:
@cabanier a depth prepass usually means rendering the entire scene once to a depth target, then binding that for compares. The article correctly notes that this is slow on mobile devices. However, here, depth would be written to the depth buffer once (usually fast) and then normal rendering happens with regular ZTest (super fast). Still think it’s worth a try!

snowzurfer commented on src/renderers/shaders/ShaderChunk/occlusion_fragment.glsl.js:
```diff
@@ -0,0 +1,20 @@
+        }
+        float assetDepthM = gl_FragCoord.z;
+        
+        //float occlusion = Depth_GetOcclusion(depthColor, depthUv, assetDepthM, arrayIndex);
```
I think this is possibly a leftover comment? maybe we could remove it

cabanier replied:
I left it here for experimentation. I suspect that we want to tweak the shader but happy to remove it.

snowzurfer reviewed (commented):
one small comment

also, I think we should use @cabanier 's current approach rather than the depth prepass one, because without it we can't have soft edges, which are quite fundamental to making the occlusion look good.

I think that most (virtually all?) AR systems have soft edges since it looks much better. Even though it introduces a bit more complexity (imo not that much), it's a worthwhile tradeoff to get soft edges.

also like @cabanier said, the two depth maps would be of different size

cabanier commented:
> @cabanier a depth prepass usually means rendering the entire scene once to a depth target, then binding that for compares. The article correctly notes that this is slow on mobile devices. However, here, depth would be written to the depth buffer once (usually fast) and then normal rendering happens with regular ZTest (super fast). Still think it’s worth a try!

It's never fast to write depth :-)
We were careful to code the browser and three.js so it never resolves depth. It's only used during the tile computation and then discarded so it never leaves the cache.

Writing regular or multisampled depth data and then reading back from it introduces too much overhead. This is also the reason why occlusion is happening at the application level and not the system compositor one and why postprocessing isn't practical.

LR17 committed:
Docs: Improve `Material` page. (#27416)

* updated Material.toneMapped docs

* Update Material.html

Clean up.

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

vanruesc committed:
Add optional timestamp parameter (#27421)

sciecode committed:
OrbitControls: pointers list only used for identification (#27420)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

sciecode committed:
OrbitControls: use scroll event delta to modulate zoom speed (#27418)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

mrdoob committed:
Timer: Added FixedTimer (#27423)

* Timer: Added FixedTimer.

* Updated docs.

* Updated docs.

mrdoob committed:
Editor: Added AgXToneMapping support.

mrdoob committed:
Examples: Change default tonemapping in webgl_tonemapping.

mrdoob committed:
Examples: Improved webgpu_compute_particles.

mrdoob committed:
Examples: Improved webgpu_backdrop_water.

mrdoob committed:
Examples: Improved webgpu_compute_particles_snow.

mrdoob committed:
Examples: Clean up.

mrdoob committed:
r160

sunag committed:
ShaderNode: Fix .isArrayInput (#27428)

hybridherbst committed:
MaterialXLoader: improve onError handling (#27433)

hybridherbst committed:
MaterialXLoader: Respect handlers attached to manager when loading textures (#27432)

Matches how GLTFloader does this

vanruesc committed:
Fix fogExp2 (#27430)

nieyuyao committed:
fix: update vertices according uv.y when linecap is square (#27429)

sciecode committed:
OrbitControls: fix Orthographic 'change' event (#27424)

* OrbitControls: fix Orthographic 'change' event

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

* OrbitControls: only recalculate when zoom has changed

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

* clean-up

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

---------

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

sciecode committed:
OrbitControls: update state on pointer up (#27425)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

Rabbid76 committed:
examples: GTAOPass fix setting of PERSPECTIVE_CAMERA (#27436)

renovate[bot] committed:
Update actions/setup-node digest to b39b52d (#27438)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action digest to 012739e (#27439)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

RenaudRohlinger committed:
WebGPURenderer: Fix alpha canvas in WebGPU (#27442)

* fix alpha background in webgpu

* fix whitespace

RenaudRohlinger committed:
GaussianBlurNode: Alpha channel support (#27444)

newstart0514 committed:
Update RectAreaLight.html (#27441)

Fix a error

vanruesc committed:
LUT Loaders: Add support for FloatType and add docs (#27431)

* Update LUT loaders

* Add docs for LUT loaders

* Fix lint issues

* Update list.json

* Add module name to error messages

* Remove unnecessary code

* Update comment

sunag committed:
WebGPURenderer: Ignore webgpu_backdrop_water from E2E tests (#27452)

RenaudRohlinger committed:
WebGPURenderer: New `AfterImageNode` for the PostProcessing renderer. (#27451)

* init afterimage node

* init value

* cleanup

---------

sunag committed:
WebGPURenderer: Fix Geometry does not have the Attribute requested by the Material (#27454)

* WebGPURenderer: Fix Geometry does not have the Attribute

* cleanup

sunag committed:
Example: `webgpu_backdrop_water` - add extra blue (#27455)

aardgoose committed:
WebGPURenderer: GaussianBlurNode reduce overhead (#27447)

* use two quadmeshes

* add tip

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

Mugen87 committed:
Updated screenshots.

hybridherbst committed:
Addons: auto-fix linting problems (#27457)

Brings 93 problems (60 errors, 33 warnings) down to 37 problems (4 errors, 33 warnings).

sciecode committed:
OrbitControls: update pointerUp (#27445)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

sciecode committed:
OrbitControls: fix pinch zoom behaviour (#27446)

Signed-off-by: Guilherme Avila <3927951+sciecode@users.noreply.github.com>

elalish committed:
Grounded Skybox (#27448)

* initial change

* mrdoob approves

* updated example

* update screenshot

mrdoob committed:
Examples: Updated webgl_materials_envmaps_groundprojected screenshot.

sunag committed:
puppeteer: add `webgpu_postprocessing_afterimage`

CodyJasonBennett committed:
Addons: export GroundedSkybox. (#27459)

Mugen87 committed:
WebGLTextures: Add filter check for float textures. (#27460)

hybridherbst committed:
MaterialX: add normal, tangent, texcoord, geomcolor, position space support (#27456)

rename to VertexColorNode

move vertex color fallback to VertexColorNode

remove unused Vector4 import

add index support to UV and color nodes

linting

another rename

Mugen87 committed:
Editor: Add compute tangents option. (#27461)

ycw committed:
Update Addons.js (#27469)

sunag committed:
MaterialXLoader: Optional parameter (#27462)

* MaterialXLoader: Optional parameter

* add some other mx_nodes

puxiao committed:
[Doc] Add some missing Chinese documents (#27443)

* Complete Chinese documents

* Update list.json

* update some

* update

* update

* update

* update

* update

* update

* update

* update

* update

* update

RenaudRohlinger committed:
WebGPURenderer: Fix Renderer `dispose()` (#27471)

* fix dispose

* temporary remove _properties

RenaudRohlinger committed:
WebGPURenderer: Forward context to Renderer (#27472)

* WebGPURenderer: WebGLBackend feed context to `Renderer`

* Forward context in WebGPU Backend too

* feedbacks

paulmasson committed:
Update LICENSE (#27476)

Mugen87 committed:
Audio: Make `setDetune()` more robust. (#27479)

aardgoose committed:
port fat line changes from standard renderer (#27474)

Co-authored-by: aardgoose <angus.sawyer@email.com>

CodyJasonBennett committed:
Constants: Remove deprecated code. (#27480)

Removes `TwoPassDoubleSide` in favor of `Material.forceSinglePass` (#25239).

RenaudRohlinger committed:
WebGPURenderer: Support KTX Compressed texture in the WebGL Backend (#27463)

* wip

* cleanup webgl backend

* cleanup and add silentError

* cleanup WebGLTextureUtils

* move textures to textureutils

* remove handling error in hasFeatre

renovate[bot] committed:
Update devDependencies (non-major) (#27481)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

gkjohnson committed:
Fix delete geometry causing the model to disappear (#27484)

Rabbid76 committed:
Examples: SVGLoader - fix triangle segment uv coordinates (#27485)

Rabbid76 committed:
examples: post-processing material AO (#27475)

* examples: post-processing material AO

* examples: postprocessing material - use gl_FragCoord.xy for map lookup

* Examples: postprocessing material ao: show diffuse pass on both sides

* Examples: postprocessing material a - simplify example

* Examples: postprocessing material ao - simplify enironment texture

WestLangley committed:
AgX: add required gamut mapping (#27413)

Mugen87 committed:
Remove deprecated code. (#27467)

* Remove deprecated code.

* Clean up.

* Examples: Bump version of `web-ifc-three`.

paulmasson committed:
Update REVISION (#27427)

sunag committed:
WebGPURenderer: MeshSSSPhysicalNodeMaterial (#27488)

* Add MeshSSSPhysicalNodeMaterial

* add `webgpu_materials_sss` example

* update title

* fix from scanning

* fix old property name

* update default values

RenaudRohlinger committed:
WebGPURenderer: MSAA, Postprocessing and Wireframe support in the WebGL Backend (#27473)

* support material.wireframe in webgl backend

* fix reattach framebuffer, support msaa and wireframe demo

* fix depth samples === 0

* cleanup

* more cleanup

* more cleanup to the webgl pipeline

* postprocessing support

* fix msaa

* fix example afterimage

* improved state management and performances

* generate a new webgl slot for fbos

* revert puppeteer

* fix copyFrameBufferToTexture with correct unbinding and cleanup

* fix multipass pipeline in webgl backend and enable new demos

* regenerate screenshot with gaussian blur

RenaudRohlinger committed:
Examples: Fix `webgpu_morphtargets_face` (#27489)

aardgoose committed:
WebGPURenderer: fix renderContext tracking in WebGLBackend (#27490)

Co-authored-by: aardgoose <angus.sawyer@email.com>

aardgoose committed:
NodeBuilder: remove duplicate code (a TODO item) (#27412)

* remove duplicate  code

* remove imports

* add switch for standard renderer path

* remove unused import

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

a0m0rajab committed:
typo fixes for installation (#27499)

* typo fixes for installation

* typo add missing word

aardgoose committed:
missing call to state helper (#27503)

Co-authored-by: aardgoose <angus.sawyer@email.com>

aardgoose committed:
WebGPURenderer: Support WebGLCubeRenderTarget (#27071)

* cube render targets

* implemenet destroyAttribute()

* replicate WebGPU texture filtering

* missed ;

* allow tests to run

* reinstate flip

* add missing ;

* repair damage

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

RenaudRohlinger committed:
MathNode: New `cbrt` Function for Easy Cube Root Calculations (#27497)

* add cbrt

* Move to TSL only

---------

Co-authored-by: sunag <sunagbrasil@gmail.com>

aardgoose committed:
WebGPURenderer: create invalidation array once for WebGL MSAA (#27505)

Co-authored-by: aardgoose <angus.sawyer@email.com>

gkjohnson committed:
Remove deprecated build files (#27507)

* Remove deprecated builds

* remove unused file

StrandedKitty committed:
Explicitly specify precision for all sampler types (#27482)

cabanier committed:
display offersession promise failure instead of showing console errors (#27492)

* display offersession promise failure instead of showing console errors

* change log() to warn() for offersession handler

Commit:
Optimize the Description of RingGeometry (#27495)

* Update RingGeometry.html

* Update RingGeometry.html

* Update RingGeometry.html

* Update RingGeometry.html

---------

Co-authored-by: Michael Herzog <michael.herzog@human-interactive.org>

Mugen87 committed:
Updated builds.

sunag committed:
Rename MeshSSSPhysicalNodeMaterial -> MeshSSSNodeMaterial (#27506)

sunag committed:
TSL: Better support for the `mod` method (#27517)

RenaudRohlinger committed:
WebGPURenderer: Remove from Puppeteer exceptionList some examples (#27514)

* test pupeeter new webgl support

* webgpu_postprocessing_afterimage seems to still get issue

* update puppeteer

RenaudRohlinger committed:
WebGPURenderer: WebGLBackend Cleanup RTT buffers when dispose texture (#27513)

* cleanup fbos

* removed unecessary cached sample as its already handled in Textures

zsombok committed:
Unify styling for default values (#27519)

susiwen8 committed:
Docs: code format and unify `three.js` (#27521)

renovate[bot] committed:
Update devDependencies (non-major) (#27524)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Sov3rain committed:
USDZExporter: Add maxTexture size option. (#27528)

RenaudRohlinger committed:
Nodes: RotateNode and remapping functions (#27525)

sunag committed:
Reduces some extra lines.

RenaudRohlinger committed:
WebGPURenderer: Add Offscreen Support (#27520)

* add offscreen support

* refactor statics webgpu

* navigator.gpu should be enough for initial check

* fix webgl context ktx2loader

* use isAvailable in webgpurenderer

* cleanup and should fix pupeeter

* return promise

* feedbacks

* remove unecessary async

* removed oversight

* cleanup

---------

aardgoose committed:
WebGPURenderer: cache renderPassDesciptors and associated views. (#27518)

* cache descriptors

* check sample count for msaa targets

* add missing ;

* set renderContext depth/stencil state correctly

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

RenaudRohlinger committed:
OrbitControls: Fix offscreen and cleanup (#27522)

* fix offscreen canvas and lint

* feedbacks

* added comment and removed if condition based on feedback

sunag committed:
WebGPUBackend: Clean up

hybridherbst committed:
GroundedSkybox: Fix UVs being flipped (#27500)

* GroundedSkybox: Fix UVs being flipped

* fix deepscan issue

Mugen87 committed:
Editor: Ensure clips are optimized before exporting to glTF. (#27529)

Mugen87 committed:
WebGLCubeMaps: Increase size of internal cube maps. (#27530)

* WebGLCubeMaps: Increase size of internal cube maps.

* Examples: Update screenshots.

Mugen87 committed:
Examples: Clean up. (#27531)

aardgoose committed:
document default (#27532)

Co-authored-by: aardgoose <angus.sawyer@email.com>

aardgoose committed:
WebGPURenderer: partial compute() shader support for WebGL backend (#27367)

* compute support

* code bot fixes

* add StorageBufferAttribute

* adapt to storagetBufferAttribute

* dynamically realign the size of storage buffer based on backend

* compute examples takes too much time to init for puppeteer

* Add StorageBufferAttribute.create()

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

gitplus committed:
Water2: Fix the error when flowSpeed and other parameters are 0. (#27536)

gitplus committed:
Editor:Fix determine if colorSpace needs to be modified (#27548)

RenaudRohlinger committed:
Nodes: AgX Tone Mapping (#27549)

* WebGPU: AgX Tone Mapping

* cleanup

liangtie committed:
fix: Update the token pattern for Identifier to load VRMLs  from Kicad (#27543)

* fix: Update the token pattern for Identifier to load VRML files from Kicad

char '-' is allowded to be used inside the identifiers for footprints in Kicad , e.g.  'PIN-01' , 'IC-LABEL-01'

* style: Replace '-' with '0x2d' in the pattern matching the identifier for vrml

* Revert "style: Replace '-' with '0x2d' in the pattern matching the identifier for vrml"

This reverts commit 11a4c67eec119cac6eb6c5a118f8ce8237a4ae37.

* Remove '0x2d' from the negation set

LeviPesin committed:
WebGLAtrributes: Add `.updateRange` deprecation comment (#27550)

tomhsiao1260 committed:
Docs: fix typo :P (#27554)

RenaudRohlinger committed:
WebGPURenderer: PostProcessing fix tone mapping on display nodes (#27557)

* disable tone mapping on postprocess quadmesh

* typo

renovate[bot] committed:
Update actions/upload-artifact digest to 1eb3cb2 (#27560)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action digest to e5f05b8 (#27561)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update dependency rollup to v4.9.5 (#27562)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

hybridherbst committed:
show which shader caused compilation error (#27553)

Mugen87 committed:
WebGLProgram: Improve formatting of error log. (#27565)

aardgoose committed:
WebGPURenderer: clear() - use cached descriptor properties and views were possible  (#27551)

* remove redundant code

* use cached descriptor properties for clear()

* remove unused import

---------

Co-authored-by: aardgoose <angus.sawyer@email.com>

hybridherbst committed:
WebXR: optional `onLoad` callback when loading hands or controllers (#27544)

* WebXR: add onLoad callback for XRHandMeshModel

* add onLoad for XRControllerModelFactory

Mugen87 committed:
Updated builds.

sunag committed:
WebGPURenderer: Fix `BachedMesh` coordinate system in case of `WebGLBackend` (#27567)

sunag committed:
Nodes: Fix `viewportCoordinate` build (#27566)

cabanier commented:
@mrdoob @Mugen87 depth sensing is now on by default in the Quest browser. Can we merge this PR?

RenaudRohlinger committed:
Nodes: Add lengthSq to MathNodes (#27568)

* add lengthSq to math nodes

* forgot lengthSq doesn't exist in glsl

* cleanup

---------

sunag committed:
WebGPURenderer: PassNode + PostProcessing - WebGPUBackend (#27537)

* WebGPURenderer: PassNode + PostProcessing

* update sky color

* cleanup

* a bit more realistic

RenaudRohlinger committed:
WebGPURenderer: Improve copyTextureToBuffer support (#27569)

* WebGPURenderer improve copyTextureToBuffer

* remove unecessary comment

* correct comment

sunag committed:
WebGPURenderer: Add renderer parameter in Textures

sunag committed:
Example webgpu_portal: cleanup

RenaudRohlinger committed:
camera proj inverse was missing (#27578)

RenaudRohlinger committed:
Nodes: Add mat4 format to `cameraProjectionMatrixInverse` (#27579)

* camera proj inverse was missing

* apply mat4 to projectionMatrixInverse

cabanier committed:
Merge branch 'occlusion' of github.com:cabanier/three.js into occlusion

hybridherbst commented:
@cabanier I was able to try this out (finally!) but still had to turn the option on on v60. Looks very jaggy/pixelated even on Quest 3 but for some experiences this might be acceptable. Is there already a newer version than v60 that has it enabled by default?

Also, some questions:
1. can the depth transition "softness" be controlled per object?
2. can the depth effect be turned off per object/per material? From the code it looks like this would just be `material.occlusion = false;`, right?

cabanier commented:
> @cabanier I was able to try this out (finally!) but still had to turn the option on on v60. Looks very jaggy/pixelated even on Quest 3 but for some experiences this might be acceptable.

True. There are better shaders than the simple ones I added. I was hoping that we experiment with them after landing the fundamentals.

>  Is there already a newer version than v60 that has it enabled by default?

It should be enabled by default in 31.1. Is that not the version you have?
OS v60 has support for depth sensing.

> 1. can the depth transition "softness" be controlled per object?

Not at the moment
> 2. can the depth effect be turned off per object/per material? From the code it looks like this would just be `material.occlusion = false;`, right?

Yes, for some content it might be desired that it's not occluded. (ie hands or the controller)

Mugen87 commented:
Closing in favor of https://github.com/mrdoob/three.js/pull/27586.

hybridherbst commented:
@Mugen87 I'm not sure if this issue should be closed, the topic of "how to allow the shader-based version in a sample" is still very much open. Many effects can't be achieved with the v2 implementation and vice versa, both have their place.

Mugen87 commented:
I'm not sure the project is willing to support both paths especially since this PR required in-depth changes in the material system. Unless @mrdoob does not decide different, `three.js` will only offer what has been implement in https://github.com/mrdoob/three.js/pull/27586. 

My hope is it will be easier to support changes like this with `WebGPURenderer`.

hybridherbst commented:
I mean like this – v2 in core, v1 as sample:
https://github.com/mrdoob/three.js/pull/27586#issuecomment-1899317734

Of course @cabanier or @mrdoob can open separate PRs for a sample that uses custom shaders, I just think it gets lost when this PR here is closed.

cabanier commented:
> I mean like this – v2 in core, v1 as sample: [#27586 (comment)](https://github.com/mrdoob/three.js/pull/27586#issuecomment-1899317734)
> 
> Of course @cabanier or @mrdoob can open separate PRs for a sample that uses custom shaders, I just think it may get lost when this PR here is closed.

I will make an attempt to extend the WebXRManager and write a test that uses a custom shader.
